{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GFF file successfully downloaded to: ./output/Arabidopsis_thaliana.TAIR10.59.chromosome.1.gff3.gz\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import requests\n",
    "from urllib.parse import urljoin\n",
    "import pandas as pd\n",
    "import gzip\n",
    "from urllib.parse import urljoin\n",
    "import os\n",
    "\n",
    "# Define the parameters\n",
    "ensembl_version = 59  # Correct release version\n",
    "organism_name = \"arabidopsis_thaliana\"\n",
    "chromosome = \"chromosome.1\"  # Change this to download specific chromosomes\n",
    "output_dir = \"./output\"\n",
    "\n",
    "# Create output directory if it doesn't exist\n",
    "os.makedirs(output_dir, exist_ok=True)\n",
    "\n",
    "",
    "def download_gff(ensembl_version: int, organism_name: str, chromosome: str, output_dir: str) -> str:\n",
    "    base_url = f\"https://ftp.ensemblgenomes.ebi.ac.uk/pub/plants/release-{ensembl_version}/gff3/{organism_name}/\"\n",
    "    gff_file = f\"{organism_name.capitalize()}.TAIR10.{ensembl_version}.{chromosome}.gff3.gz\"\n",
    "    full_url = urljoin(base_url, gff_file)\n",
    "    local_path = os.path.join(output_dir, gff_file)\n",
    "\n",
    "    # Download the file\n",
    "    response = requests.get(full_url)\n",
    "    if response.status_code == 200:\n",
    "        with open(local_path, 'wb') as f:\n",
    "            f.write(response.content)\n",
    "        return local_path\n",
    "    else:\n",
    "        raise Exception(f\"Failed to download GFF file for {organism_name} {chromosome} (status code: {response.status_code})\")\n",
    "# Test the download with chromosome 1\n",
    "try:\n",
    "    gff_file_path = download_gff(ensembl_version, organism_name, chromosome, output_dir)\n",
    "    print(f\"GFF file successfully downloaded to: {gff_file_path}\")\n",
    "except Exception as e:\n",
    "    print(f\"Error during download: {e}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the parse_gff function\n",
    "def parse_gff(file_path: str) -> pd.DataFrame:\n",
    "    with gzip.open(file_path, 'rt') as f:\n",
    "        gff_data = pd.read_csv(f, sep='\\t', comment='#', header=None, names=[\n",
    "            'seqid', 'source', 'feature', 'start', 'end', 'score', 'strand', 'phase', 'attributes'\n",
    "        ])\n",
    "    \n",
    "    genes = gff_data[gff_data['feature'] == 'gene']\n",
    "\n",
    "    def parse_attributes(attributes):\n",
    "        attributes_dict = dict(attr.split('=') for attr in attributes.split(';'))\n",
    "        return attributes_dict.get('ID'), attributes_dict.get('Name')\n",
    "\n",
    "    genes[['gene_id', 'gene_name']] = genes['attributes'].apply(lambda x: pd.Series(parse_attributes(x)))\n",
    "\n",
    "    return genes[['seqid', 'start', 'end', 'gene_id', 'gene_name', 'strand']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set parameters\n",
    "ensembl_version = 59\n",
    "organism_name = \"arabidopsis_thaliana\"\n",
    "chromosome = \"chromosome.1\"\n",
    "output_dir = \"./output\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gff_file_path = download_gff(ensembl_version, organism_name, chromosome, output_dir)\n",
    "print(f\"GFF file successfully downloaded to: {gff_file_path}\")\n",
    "genes_df = parse_gff(gff_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "DataFrame Info:\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Index: 7156 entries, 1 to 213201\n",
      "Data columns (total 6 columns):\n",
      " #   Column     Non-Null Count  Dtype \n",
      "---  ------     --------------  ----- \n",
      " 0   seqid      7156 non-null   int64 \n",
      " 1   start      7156 non-null   int64 \n",
      " 2   end        7156 non-null   int64 \n",
      " 3   gene_id    7156 non-null   object\n",
      " 4   gene_name  7140 non-null   object\n",
      " 5   strand     7156 non-null   object\n",
      "dtypes: int64(3), object(3)\n",
      "memory usage: 391.3+ KB\n",
      "None\n",
      "\n",
      "First few rows of the DataFrame:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>3631</td>\n",
       "      <td>5899</td>\n",
       "      <td>gene:AT1G01010</td>\n",
       "      <td>NAC001</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1</td>\n",
       "      <td>6788</td>\n",
       "      <td>9130</td>\n",
       "      <td>gene:AT1G01020</td>\n",
       "      <td>ARV1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>1</td>\n",
       "      <td>11649</td>\n",
       "      <td>13714</td>\n",
       "      <td>gene:AT1G01030</td>\n",
       "      <td>NGA3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>1</td>\n",
       "      <td>23121</td>\n",
       "      <td>31227</td>\n",
       "      <td>gene:AT1G01040</td>\n",
       "      <td>DCL1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245</th>\n",
       "      <td>1</td>\n",
       "      <td>31170</td>\n",
       "      <td>33171</td>\n",
       "      <td>gene:AT1G01050</td>\n",
       "      <td>PPa1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     seqid  start    end         gene_id gene_name strand\n",
       "1        1   3631   5899  gene:AT1G01010    NAC001      +\n",
       "17       1   6788   9130  gene:AT1G01020      ARV1      -\n",
       "135      1  11649  13714  gene:AT1G01030      NGA3      -\n",
       "152      1  23121  31227  gene:AT1G01040      DCL1      +\n",
       "245      1  31170  33171  gene:AT1G01050      PPa1      -"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "DataFrame Description:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>7156.0</td>\n",
       "      <td>7.156000e+03</td>\n",
       "      <td>7.156000e+03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.494342e+07</td>\n",
       "      <td>1.494586e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>0.0</td>\n",
       "      <td>9.528129e+06</td>\n",
       "      <td>9.528120e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>1.0</td>\n",
       "      <td>3.631000e+03</td>\n",
       "      <td>5.899000e+03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>1.0</td>\n",
       "      <td>6.066435e+06</td>\n",
       "      <td>6.068832e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.402051e+07</td>\n",
       "      <td>1.402270e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>1.0</td>\n",
       "      <td>2.373760e+07</td>\n",
       "      <td>2.373907e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>1.0</td>\n",
       "      <td>3.042442e+07</td>\n",
       "      <td>3.042519e+07</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqid         start           end\n",
       "count  7156.0  7.156000e+03  7.156000e+03\n",
       "mean      1.0  1.494342e+07  1.494586e+07\n",
       "std       0.0  9.528129e+06  9.528120e+06\n",
       "min       1.0  3.631000e+03  5.899000e+03\n",
       "25%       1.0  6.066435e+06  6.068832e+06\n",
       "50%       1.0  1.402051e+07  1.402270e+07\n",
       "75%       1.0  2.373760e+07  2.373907e+07\n",
       "max       1.0  3.042442e+07  3.042519e+07"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Missing Values:\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "seqid         0\n",
       "start         0\n",
       "end           0\n",
       "gene_id       0\n",
       "gene_name    16\n",
       "strand        0\n",
       "dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Top 10 Longest Genes:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>strand</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>114296</th>\n",
       "      <td>1</td>\n",
       "      <td>17732010</td>\n",
       "      <td>17758508</td>\n",
       "      <td>gene:AT1G48090</td>\n",
       "      <td>AT1G48090</td>\n",
       "      <td>-</td>\n",
       "      <td>26498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169043</th>\n",
       "      <td>1</td>\n",
       "      <td>25069428</td>\n",
       "      <td>25095638</td>\n",
       "      <td>gene:AT1G67120</td>\n",
       "      <td>AT1G67120</td>\n",
       "      <td>-</td>\n",
       "      <td>26210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146678</th>\n",
       "      <td>1</td>\n",
       "      <td>21746354</td>\n",
       "      <td>21766048</td>\n",
       "      <td>gene:AT1G58602</td>\n",
       "      <td>AT1G58602</td>\n",
       "      <td>+</td>\n",
       "      <td>19694</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163088</th>\n",
       "      <td>1</td>\n",
       "      <td>24064822</td>\n",
       "      <td>24083428</td>\n",
       "      <td>gene:AT1G64790</td>\n",
       "      <td>ILA</td>\n",
       "      <td>-</td>\n",
       "      <td>18606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26374</th>\n",
       "      <td>1</td>\n",
       "      <td>2588854</td>\n",
       "      <td>2606892</td>\n",
       "      <td>gene:AT1G08260</td>\n",
       "      <td>TIL1</td>\n",
       "      <td>+</td>\n",
       "      <td>18038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121761</th>\n",
       "      <td>1</td>\n",
       "      <td>18522248</td>\n",
       "      <td>18539995</td>\n",
       "      <td>gene:AT1G50030</td>\n",
       "      <td>TOR</td>\n",
       "      <td>-</td>\n",
       "      <td>17747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169357</th>\n",
       "      <td>1</td>\n",
       "      <td>25100848</td>\n",
       "      <td>25117531</td>\n",
       "      <td>gene:AT1G67140</td>\n",
       "      <td>SWEETIE</td>\n",
       "      <td>-</td>\n",
       "      <td>16683</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7844</th>\n",
       "      <td>1</td>\n",
       "      <td>712440</td>\n",
       "      <td>727553</td>\n",
       "      <td>gene:AT1G03060</td>\n",
       "      <td>SPI</td>\n",
       "      <td>-</td>\n",
       "      <td>15113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145817</th>\n",
       "      <td>1</td>\n",
       "      <td>21587060</td>\n",
       "      <td>21601740</td>\n",
       "      <td>gene:AT1G58250</td>\n",
       "      <td>SAB</td>\n",
       "      <td>-</td>\n",
       "      <td>14680</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205734</th>\n",
       "      <td>1</td>\n",
       "      <td>29818970</td>\n",
       "      <td>29833016</td>\n",
       "      <td>gene:AT1G79280</td>\n",
       "      <td>NUA</td>\n",
       "      <td>-</td>\n",
       "      <td>14046</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqid     start       end         gene_id  gene_name strand  length\n",
       "114296      1  17732010  17758508  gene:AT1G48090  AT1G48090      -   26498\n",
       "169043      1  25069428  25095638  gene:AT1G67120  AT1G67120      -   26210\n",
       "146678      1  21746354  21766048  gene:AT1G58602  AT1G58602      +   19694\n",
       "163088      1  24064822  24083428  gene:AT1G64790        ILA      -   18606\n",
       "26374       1   2588854   2606892  gene:AT1G08260       TIL1      +   18038\n",
       "121761      1  18522248  18539995  gene:AT1G50030        TOR      -   17747\n",
       "169357      1  25100848  25117531  gene:AT1G67140    SWEETIE      -   16683\n",
       "7844        1    712440    727553  gene:AT1G03060        SPI      -   15113\n",
       "145817      1  21587060  21601740  gene:AT1G58250        SAB      -   14680\n",
       "205734      1  29818970  29833016  gene:AT1G79280        NUA      -   14046"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Number of duplicate gene IDs: 0\n",
      "\n",
      "Number of genes without names: 16\n",
      "Sample of Unnamed Genes:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>strand</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>23284</th>\n",
       "      <td>1</td>\n",
       "      <td>2295553</td>\n",
       "      <td>2296497</td>\n",
       "      <td>gene:AT1G07476</td>\n",
       "      <td>None</td>\n",
       "      <td>+</td>\n",
       "      <td>944</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55995</th>\n",
       "      <td>1</td>\n",
       "      <td>5812728</td>\n",
       "      <td>5816662</td>\n",
       "      <td>gene:AT1G17000</td>\n",
       "      <td>None</td>\n",
       "      <td>+</td>\n",
       "      <td>3934</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60323</th>\n",
       "      <td>1</td>\n",
       "      <td>6304360</td>\n",
       "      <td>6305689</td>\n",
       "      <td>gene:AT1G18320</td>\n",
       "      <td>None</td>\n",
       "      <td>+</td>\n",
       "      <td>1329</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62926</th>\n",
       "      <td>1</td>\n",
       "      <td>6589835</td>\n",
       "      <td>6592762</td>\n",
       "      <td>gene:AT1G19090</td>\n",
       "      <td>None</td>\n",
       "      <td>+</td>\n",
       "      <td>2927</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78450</th>\n",
       "      <td>1</td>\n",
       "      <td>8692630</td>\n",
       "      <td>8694705</td>\n",
       "      <td>gene:AT1G24530</td>\n",
       "      <td>None</td>\n",
       "      <td>+</td>\n",
       "      <td>2075</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqid    start      end         gene_id gene_name strand  length\n",
       "23284      1  2295553  2296497  gene:AT1G07476      None      +     944\n",
       "55995      1  5812728  5816662  gene:AT1G17000      None      +    3934\n",
       "60323      1  6304360  6305689  gene:AT1G18320      None      +    1329\n",
       "62926      1  6589835  6592762  gene:AT1G19090      None      +    2927\n",
       "78450      1  8692630  8694705  gene:AT1G24530      None      +    2075"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Explore the DataFrame\n",
    "# Display basic information about the DataFrame\n",
    "print(\"\\nDataFrame Info:\")\n",
    "print(genes_df.info())\n",
    "\n",
    "print(\"\\nFirst few rows of the DataFrame:\")\n",
    "display(genes_df.head())\n",
    "\n",
    "print(\"\\nDataFrame Description:\")\n",
    "display(genes_df.describe())\n",
    "\n",
    "# Check for missing values\n",
    "print(\"\\nMissing Values:\")\n",
    "display(genes_df.isnull().sum())\n",
    "\n",
    "# Gene lengths\n",
    "genes_df['length'] = genes_df['end'] - genes_df['start']\n",
    "\n",
    "# Top 10 longest genes\n",
    "print(\"\\nTop 10 Longest Genes:\")\n",
    "display(genes_df.nlargest(10, 'length'))\n",
    "\n",
    "# Check for duplicate gene IDs\n",
    "duplicate_ids = genes_df[genes_df.duplicated(subset='gene_id', keep=False)]\n",
    "print(f\"\\nNumber of duplicate gene IDs: {len(duplicate_ids)}\")\n",
    "if len(duplicate_ids) > 0:\n",
    "    print(\"Duplicate Gene IDs:\")\n",
    "    display(duplicate_ids)\n",
    "\n",
    "# Check for genes without names\n",
    "unnamed_genes = genes_df[genes_df['gene_name'].isnull()]\n",
    "print(f\"\\nNumber of genes without names: {len(unnamed_genes)}\")\n",
    "if len(unnamed_genes) > 0:\n",
    "    print(\"Sample of Unnamed Genes:\")\n",
    "    display(unnamed_genes.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
